edac: Add Error Detection And Correction (EDAC) and IBECC driver

finikorg opened the pull request:
Add IBECC driver, EDAC API, EDAC shell and tests.

Tested with the provided tests on ehl_crb platform with following parameters:
```
$ zephyr/scripts/twister -p ehl_crb --device-testing --device-serial-pty=<my console script> --west-flash=<my flash script> -t edac -v
```
with following results
```
INFO    - Zephyr version: zephyr-v2.4.0-2557-g6707d4822d
INFO    - JOBS: 12
INFO    - Building initial testcase list...
INFO    - 1057 test suites (1030 configurations) selected, 1027 configurations discarded due to filters.

Device testing on:

| Platform   | ID   | Serial device   |
|------------|------|-----------------|
| ehl_crb    |      |                 |

INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - 1/3 ehl_crb                   tests/subsys/edac/ibecc/edac.ibecc.debug           PASSED (device 33.825s)
INFO    - 2/3 ehl_crb                   tests/subsys/edac/ibecc/edac.ibecc.production      PASSED (device 31.222s)
INFO    - 3/3 ehl_crb                   samples/subsys/edac/sample.subsys.edac             PASSED (device 31.283s)

INFO    - 3 of 3 test configurations passed (100.00%), 0 failed, 1027 skipped with 0 warnings in 227.43 seconds
INFO    - In total 7 test cases were executed, 9179 skipped on 1 out of total 314 platforms (0.32%)
INFO    - 3 test configurations executed on platforms, 0 test configurations were only built.

Hardware distribution summary:

| Board   | ID   |   Counter |
|---------|------|-----------|
| ehl_crb |      |         3 |
```

finikorg committed:
edac: tests: Add testing EDAC IBECC API

Add testing EDAC API, can be run on device ehl_crb.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: Add EDAC API

Add Error Detection And Correction (EDAC) API.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: tests: Add injection API test

Add test injecting error and getting correct notification about EDAC
error. Since we have logging enabled in the exception handler it does
not work with LOG_MINIMAL.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: Add Zephyr EDAC configuration

Add Kconfig for EDAC.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: edac: Add EDAC shell sample application

Add EDAC shell application demonstrating EDAC API usage. Can be used
for verification of the functionality with error injection mechanism.
Sample is also included in the twister tests.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: shell: Add EDAC shell

Add EDAC shell module which would be included in the EDAC sample.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: ibecc: Add EDAC IBECC driver

Add In-Band ECC driver. Supported in ehl_crb board.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jhedberg commented on drivers/edac/edac_ibecc.c:
```diff
@@ -0,0 +1,454 @@
+	/* Memory configuration */
+
+	chan_hash =  sys_read32(HOST_REG(data->mchbar, CHANNEL_HASH));
+	printk("Channel hash %x\n", chan_hash);
```
I guess you should be using LOG_* when you even declare your own log module in the beginning of this file?

finikorg replied:
I replaced all printk to LOG_DBG()

jhedberg commented on drivers/edac/edac_ibecc.c:
```diff
@@ -0,0 +1,454 @@
+	printk("Channel hash %x\n", chan_hash);
+
+	mad_inter_chan = sys_read32(HOST_REG(data->mchbar, MAD_INTER_CHAN));
+	printk("DDR memory type %s\n",
+	       get_ddr_type(INTER_CHAN_DDR_TYPE(mad_inter_chan)));
```
Same here

jukkar commented on include/edac.h:
```diff
@@ -0,0 +1,255 @@
+
+#include <sys/types.h>
+
+typedef int (*edac_api_inject_addr_set)(const struct device *dev,
+					uint64_t addr);
+typedef uint64_t (*edac_api_inject_addr_get)(const struct device *dev);
+
+typedef int (*edac_api_inject_addr_mask_set)(const struct device *dev,
+					     uint64_t mask);
+typedef uint64_t (*edac_api_inject_addr_mask_get)(const struct device *dev);
+
+typedef int (*edac_api_inject_ctrl_set)(const struct device *dev, uint32_t ctrl);
+
+typedef uint64_t (*edac_api_ecc_error_log_get)(const struct device *dev);
+typedef void (*edac_api_ecc_error_log_clear)(const struct device *dev);
+
+typedef uint64_t (*edac_api_parity_error_log_get)(const struct device *dev);
+typedef void (*edac_api_parity_error_log_clear)(const struct device *dev);
+
+typedef unsigned int (*edac_api_errors_cor_get)(const struct device *dev);
+typedef unsigned int (*edac_api_errors_uc_get)(const struct device *dev);
+
+typedef void (*edac_notify_callback_t)(const struct device *dev,
+				       void *data);
+typedef int (*edac_api_notify_cb_set)(const struct device *dev,
+				      edac_notify_callback_t cb);
```
Should we follow some naming convention for the function typedef, Tomasz used `_f` prefix for ivshmem PR, should we follow suit?

finikorg replied:
added `_f` suffix.

jukkar commented on include/edac.h:
```diff
@@ -0,0 +1,255 @@
+/*
```
Should we put the header to some subdirectory, perhaps drivers? Your have `ZEPHYR_INCLUDE_DRIVERS_EDAC_H_` already mentioning drivers.

finikorg replied:
moved to `include/drivers/`

jukkar commented on drivers/edac/shell.c:
```diff
@@ -0,0 +1,389 @@
+/*
```
You are missing commit body for "edac: shell: .." commit.

finikorg replied:
fixed

finikorg committed:
edac: tests: Include sample.subsys.edac only for ehl_crb

Fixes twister build.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
CODEOWNERS: Adding drivers/edac owner

Adding myself to the list for drivers/edac.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
I am going to use PCI API from #29867 when it is merged.

martin-aberg commented:
Having an EDAC API in Zephyr is a great thing and the shell command is an interesting user interface.

Concern: the EDAC API as proposed in this PR appears to be heavily tailored for one specific controller.

It would be useful for the _Error Injection API_ to have abstractions along the lines of _"inject one single-bit error on `bit` at address `addr`"_. That would make the Zephyr EDAC API independent of the details of the EDAC controller registers (which may not share the notion of `addr`, `mask` and `ctrl`).

For the _Error Logging API_, consider providing a way to describe errors in more detail: how and when an error was detected (scrubber, cache fetch, DMA read, which bus master, etc).

finikorg commented:
> Concern: the EDAC API as proposed in this PR appears to be heavily tailored for one specific controller.
> 
> It would be useful for the _Error Injection API_ to have abstractions along the lines of _"inject one single-bit error on `bit` at address `addr`"_. That would make the Zephyr EDAC API independent of the details of the EDAC controller registers (which may not share the notion of `addr`, `mask` and `ctrl`).

I do not know how other platforms implements error injection, I just checked Linux
https://www.kernel.org/doc/Documentation/acpi/apei/einj.txt
There is error_type, parameters: param1, param2, param3, flags  and error_inject which triggers injection.
So I can keep address and mask as param1 and param2 and other platforms may put there whatever they want.
This would rename `edac_inject_addr_set()` to `edac_inject_set_param1()`, etc.
Then I can implement set and get error_type.

> 
> For the _Error Logging API_, consider providing a way to describe errors in more detail: how and when an error was detected (scrubber, cache fetch, DMA read, which bus master, etc).

I am not sure you can get this information. BTW: How are you going to use it?

finikorg committed:
edac: Generalize EDAC API

Replace addr with param1 and addr_mask with param2 for get / set types
of functions. Those names are more general and allow to implement
error injections for other platforms.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

martin-aberg commented:
Thank you for the clarification and the Linux reference. My updated understanding is that the PR provides an API for the ACPI APEI specification.

_EDAC_ is a rather broad concept. In some computing domains (space) you will find EDAC used anywhere you may have a memory, including  external RAM, caches, CPU register file, network equipment routing tables, etc. [Here is an example](https://www.gaisler.com/doc/gr712rc-usermanual.pdf#G8.966345) of a memory controller interface which can do injection and count errors, but is not ACPI APEI.

That being said, I see no problem with the API being specific to the ACPI APEI. But you could consider if you want to rename it to "APEI API". Either way, it would be helpful if APEI was at least mentioned in the API documentation and/or the Kconfig.

finikorg committed:
edac: Split EDAC set ctrl to set error_type and error_trigger

Split setting EDAC IBECC ctrl to setting error_type and
error_trigger to make it easier for other platforms.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
> Thank you for the clarification and the Linux reference. My updated understanding is that the PR provides an API for the ACPI APEI specification.
> 

It is not, I just used something similar to accommodate for the range of controller.

> _EDAC_ is a rather broad concept. In some computing domains (space) you will find EDAC used anywhere you may have a memory, including external RAM, caches, CPU register file, network equipment routing tables, etc. [Here is an example](https://www.gaisler.com/doc/gr712rc-usermanual.pdf#G8.966345) of a memory controller interface which can do injection and count errors, but is not ACPI APEI.

So what is the software API for that type of controllers? I see only register description.

> That being said, I see no problem with the API being specific to the ACPI APEI. But you could consider if you want to rename it to "APEI API". Either way, it would be helpful if APEI was at least mentioned in the API documentation and/or the Kconfig.

Since I use only some elements of APEI I do not think we might use this name.

finikorg committed:
edac: Use helpers to read / write IBECC registers

Use helpers ibecc_{read,write}_reg{32,64}.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: Remove extra logging

Remove unneeded extra logging.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: Add error data information to callback parameter

Add extra error data information to callback parameter. Add tests for
testing the data provided.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: Replace DEVICE_AND_API_INIT with DEVICE_DEFINE

Since DEVICE_AND_API_INIT was recently deprecated replace with correct
version.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: doc: Include EDAC API to Zephyr API Reference

Include EDAC API to generated Zephyr API Reference.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

tbursztyka commented on drivers/edac/edac_ibecc.c:
```diff
@@ -0,0 +1,457 @@
+
+	tolud = pcie_conf_read(bdf, TOLUD_REG) & GENMASK(31, 20);
+
+	device_map(&data->mchbar, mchbar, 0x10000, K_MEM_CACHE_NONE);
```
What is the reason for not using pcie_get_mbar() ? 
The value 0x10000 comes from the datasheet? (pcie_get_mbar() would give you that info btw).

finikorg replied:
MCHBAR is base address for the Host Memory Mapped Configuration space, at 0x48.

nashif replied:
You should avoid magic numbers and declare this with details about where this number came from.

finikorg replied:
I will fix this

tbursztyka commented on drivers/edac/edac_ibecc.c:
```diff
@@ -0,0 +1,457 @@
+	uint32_t tolud;
+	uint64_t touud, tom, mchbar;
+
+	ARG_UNUSED(dev);
```
useless, dev is beitg used above to retrieve the private data.

finikorg replied:
I will remove this.

finikorg commented:
Rebased, since PR had merge conflict.

jhedberg reviewed (approved):
Looks ok, but please take into account the comment from @nashif about magic numbers, and you should probably next look into making this properly use Device Tree (unless I missed it, you've got hard-coded assumptions about the HW rather than having a node in DTS and using a proper DT binding for it).

jhedberg commented on drivers/edac/ibecc.h:
```diff
@@ -0,0 +1,159 @@
+/* TODO: Move to correct place NMI registers */
+
+/* NMI Status and Control Register (NMI_STS_CNT) */
+#define NMI_STS_CNT_REG		0x61
```
Minor thing, but applies to everywhere where you have defines etc: please only use tabs for indentation and never for alignment (i.e. only spaces should be used for that). Here you've used tabs for alignment which makes this render inconsistently e.g. in a browser when reviewing using the GitHub web interface.

jenmwms replied:
> please take into account the comment from @nashif about magic numbers, and you should probably next look into making this properly use Device Tree

Perhaps this could be an enhancement?

nashif replied:
if it is not fixed in this PR, it will just get lost. Something like that is not worth opening an enhancement issue for.

finikorg replied:
> Minor thing, but applies to everywhere where you have defines etc: please only use tabs for indentation and never for alignment (i.e. only spaces should be used for that). Here you've used tabs for alignment which makes this render inconsistently e.g. in a browser when reviewing using the GitHub web interface.

Is this mentioned in the Coding Style document? I think I usually see in headers alignments by tabs, like in include/drivers/pcie/pcie.h, etc.

finikorg replied:
@nashif @jhedberg @jenmwms I can make driver use Device Tree, just need to invent what kind of useless information I am going to get from it ;)

jhedberg replied:
> Is this mentioned in the Coding Style document? I think I usually see in headers alignments by tabs, like in include/drivers/pcie/pcie.h, etc.

I'm not sure if it's mentioned explicitly anywhere. And yes, the Zephyr tree is wildly inconsistent with this. However, what I've found is that "tabs for indentation, spaces for alignment" is what gives the most readable result in multiple UIs (both editors and web browsers).

finikorg replied:
I have created simple DTS: 5735ba12c0e7f38dd342cb82ba71d864de960e8e. I pass BDF and Intel Vendor ID, well known and harmful identifiers. For the product ID we might need to pass list of IDs.

nashif replied:
> @nashif @jhedberg @jenmwms I can make driver use Device Tree, just need to invent what kind of useless information I am going to get from it ;)

I am not referring to DT, just the one magic number and the need to define it somewhere with details about where it came from.

finikorg replied:
@nashif I will fix this, there might be some other magic numbers like `GENMASK(31, 20)`, I might fix it also.

finikorg replied:
Added 06c874712575025ee5f767936c3e11d27437b20e

jenmwms commented:
> I am going to use PCI API from #29867 when it is merged.

@finikorg Does #29867 need to be merged first?

finikorg commented:
> > I am going to use PCI API from #29867 when it is merged.
> 
> @finikorg Does #29867 need to be merged first?

No, At the moment I am using current ways of reading PCI.

finikorg committed:
edac: Add DTS bindings for IBECC

Add bindings for EDAC/IBECC. At the moment we can pass BDF and Vendor
ID.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
edac: Remove magic numbers

Use proper defines instead of magic numbers.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>
